File: bots/FloodFillBot.cpp
/*
 * FloodFill Bot (C++)
 * ====================
 * 策略：每步枚举 4 个方向，选走后生存空间（BFS 可达格数）最大的方向。
 *
 * 上传到 KOB 平台时，将本文件全部内容粘贴到代码框中即可。
 * 语言选择：C++
 * 编译：g++ -O2 -std=c++17 -o bot FloodFillBot.cpp
 */

#include <bits/stdc++.h>
using namespace std;

const int ROWS = 13, COLS = 14;
const int DX[] = {-1, 0, 1, 0};
const int DY[] = {0,  1, 0, -1};

/**
 * Determines whether the snake tail should grow at the given step.
 * 判断在给定步数时蛇尾是否应增长。
 */
bool checkTailIncreasing(int step) {
    if (step <= 10) return true;
    return step % 3 == 1;
}

/**
 * Rebuilds snake body cells from start position and step history.
 * 根据起点和历史步骤重建蛇身格子。
 */
vector<pair<int,int>> getCells(int sx, int sy, const string& steps) {
    deque<pair<int,int>> cells;
    cells.push_back({sx, sy});
    int x = sx, y = sy, step = 0;
    for (char c : steps) {
        int d = c - '0';
        x += DX[d]; y += DY[d];
        cells.push_back({x, y});
        step++;
        if (!checkTailIncreasing(step)) cells.pop_front();
    }
    return {cells.begin(), cells.end()};
}

/**
 * Runs BFS and returns reachable free-cell count from the given start cell.
 * 运行BFS并返回从给定起点可到达的空白格数量。
 */
int floodFill(int g[ROWS][COLS], int sx, int sy) {
    if (g[sx][sy] != 0) return 0;
    bool visited[ROWS][COLS] = {};
    queue<pair<int,int>> q;
    q.push({sx, sy});
    visited[sx][sy] = true;
    int count = 1;
    while (!q.empty()) {
        auto [cx, cy] = q.front(); q.pop();
        for (int d = 0; d < 4; d++) {
            int nx = cx + DX[d], ny = cy + DY[d];
            if (nx >= 0 && nx < ROWS && ny >= 0 && ny < COLS
                    && g[nx][ny] == 0 && !visited[nx][ny]) {
                visited[nx][ny] = true;
                count++;
                q.push({nx, ny});
            }
        }
    }
    return count;
}

/**
 * Parses input, evaluates four directions, and outputs the safest move direction.
 * 解析输入、评估四个方向并输出最安全的移动方向。
 */
int main() {
    // ── 读取输入 ──────────────────────────────────────────────────────────────
    ifstream fin("input.txt");
    string data; fin >> data; fin.close();

    // 按 '#' 分割
    vector<string> parts;
    {
        stringstream ss(data);
        string token;
        while (getline(ss, token, '#')) parts.push_back(token);
    }

    string mapStr  = parts[0];
    int meSx  = stoi(parts[1]), meSy  = stoi(parts[2]);
    string meSteps  = parts[3].substr(1, parts[3].size() - 2);   // 去掉括号
    int youSx = stoi(parts[4]), youSy = stoi(parts[5]);
    string youSteps = parts[6].substr(1, parts[6].size() - 2);

    // ── 建图 ──────────────────────────────────────────────────────────────────
    int g[ROWS][COLS] = {};
    for (int i = 0; i < ROWS; i++)
        for (int j = 0; j < COLS; j++)
            if (mapStr[i * COLS + j] == '1') g[i][j] = 1;

    auto meCells  = getCells(meSx,  meSy,  meSteps);
    auto youCells = getCells(youSx, youSy, youSteps);
    for (auto& c : meCells)  g[c.first][c.second] = 1;
    for (auto& c : youCells) g[c.first][c.second] = 1;

    // ── 决策 ──────────────────────────────────────────────────────────────────
    auto myHead = meCells.back();
    int bestDir = 0, bestSpace = -1;

    for (int d = 0; d < 4; d++) {
        int nx = myHead.first  + DX[d];
        int ny = myHead.second + DY[d];
        if (nx < 0 || nx >= ROWS || ny < 0 || ny >= COLS) continue;
        if (g[nx][ny] != 0) continue;
        int space = floodFill(g, nx, ny);
        if (space > bestSpace) {
            bestSpace = space;
            bestDir   = d;
        }
    }

    cout << bestDir << endl;
    return 0;
}

File: bots/FloodFillBot.py
"""
FloodFill Bot (Python)
======================
策略：每步枚举 4 个方向，选走后生存空间（BFS 可达格数）最大的方向。

上传到 KOB 平台时，将本文件全部内容粘贴到代码框中即可。
语言选择：Python
"""
from collections import deque

ROWS, COLS = 13, 14
DX = [-1, 0, 1, 0]
DY = [0, 1, 0, -1]


# 判断在给定步数时蛇尾是否应增长。
# Determines whether the snake tail should grow at the given step.
def check_tail_increasing(step: int) -> bool:
    if step <= 10:
        return True
    return step % 3 == 1


# 根据起点和历史步骤重建蛇身格子。
# Rebuilds snake body cells from start position and step history.
def get_cells(sx: int, sy: int, steps: str):
    cells = [[sx, sy]]
    x, y, step = sx, sy, 0
    for c in steps:
        d = int(c)
        x += DX[d]
        y += DY[d]
        cells.append([x, y])
        step += 1
        if not check_tail_increasing(step):
            cells.pop(0)
    return cells


# 运行BFS并返回从给定起点可到达的空白格数量。
# Runs BFS and returns reachable free-cell count from the given start cell.
def flood_fill(g, sx: int, sy: int) -> int:
    if g[sx][sy] != 0:
        return 0
    queue = deque([[sx, sy]])
    visited = [[False] * COLS for _ in range(ROWS)]
    visited[sx][sy] = True
    count = 1
    while queue:
        cx, cy = queue.popleft()
        for d in range(4):
            nx, ny = cx + DX[d], cy + DY[d]
            if 0 <= nx < ROWS and 0 <= ny < COLS and g[nx][ny] == 0 and not visited[nx][ny]:
                visited[nx][ny] = True
                count += 1
                queue.append([nx, ny])
    return count


# ── 读取输入 ──────────────────────────────────────────────────────────────────
with open("input.txt") as f:
    data = f.read().strip()

parts = data.split("#")
map_str  = parts[0]
me_sx,  me_sy  = int(parts[1]), int(parts[2])
me_steps       = parts[3][1:-1]   # 去掉首尾括号
you_sx, you_sy = int(parts[4]), int(parts[5])
you_steps      = parts[6][1:-1]

# ── 建图 ──────────────────────────────────────────────────────────────────────
g = [[0] * COLS for _ in range(ROWS)]
for i in range(ROWS):
    for j in range(COLS):
        if map_str[i * COLS + j] == "1":
            g[i][j] = 1

me_cells  = get_cells(me_sx,  me_sy,  me_steps)
you_cells = get_cells(you_sx, you_sy, you_steps)
for c in me_cells:  g[c[0]][c[1]] = 1
for c in you_cells: g[c[0]][c[1]] = 1

# ── 决策 ──────────────────────────────────────────────────────────────────────
my_head = me_cells[-1]
best_dir, best_space = 0, -1

for d in range(4):
    nx, ny = my_head[0] + DX[d], my_head[1] + DY[d]
    if nx < 0 or nx >= ROWS or ny < 0 or ny >= COLS:
        continue
    if g[nx][ny] != 0:
        continue
    space = flood_fill(g, nx, ny)
    if space > best_space:
        best_space = space
        best_dir = d

print(best_dir)
